### DISCUSSION TITLE
2019 August dosemu2 crashes with 2021 gcc FreeDOS kernel

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1549] [General]
I used to use dosemu2 `2.0pre8-1234-g0984731c9` of 2019 August as my default dosemu2 until today. I found that with the FreeDOS kernel I built with ia16 gcc from [the current revision](https://github.com/FDOS/kernel/commit/cedcaee5adbc2d0d4d08c3572aae8decf50d4bb4) the machine would crash once emufs.com executes [its first dosemu helper interrupt](https://github.com/dosemu2/dosemu2/blob/0984731c9b8730b0b54c55360943522c7d0c86db/src/plugin/doscmd/generic.s#L28-L30). When using lDebug as a bootloaded debugger this caused various silent or loud crashes. Without the debugger the kernel would boot a second time at that point, and pop up a question on whether to boot off a floppy diskette or the hard disk, with an 18s timeout. Selecting hard disk booted through config.sys and autoexec.bat again but with more errors. Selecting diskette crashed more quickly.

I eventually found that [a 2019 November commit](https://github.com/dosemu2/dosemu2/commit/9cf2b756098eb20f787670cdb7e560af53e2f3a5) seems to be the first one that works. I found out about it using some automated and some manual bisecting, using [this script](https://pushbx.org/ecm/test/20211031/test3.sh).

The kernel I used previously (build 2042 ish, Borland compiled) appears to work with all of the dosemu2 revisions I have used. Presumably it isn't affected by the register changes that were fixed by the 2019 November commit.

**Comment by stsp** [comment]
Not sure what is your motivation
for using such an old software.

**Reply by ecm-pushbx** [reply]
> Not sure what is your motivation for using such an old software.

I prefer using something that is somewhat stable. I recently found a bug due to [an addition you made to int.c](https://github.com/dosemu2/dosemu2/blob/d7402eec84478c051d25e7b26dd8515514c186e2/src/base/core/int.c#L1633-L1639) which made my TSR applications' installers fail. They assumed that for process relocation they could set up the interrupt 21h service 4Ch user stack frame so that the kernel's termination (terminate old process, transfer control to new process as if it was the parent) would exactly re-use this frame (pointed to by parent's `dword [PSP:2Eh]`) as if the parent had called DOS (for say a service 4Bh call) before. So that all registers are preserved across the DOS 4Ch call.

The ones I actually noticed happened to work mostly okay, it just happened sometimes that the random value that `bp` received had the /J switch's bit set and consequently the resident would be installed to the Low Memory Area. (My TSRs will show a message like "high loaded" as opposed to just "loaded" to indicate where they installed the resident.)

However, I recently found that KEEPHOOK fails rather more spectacularly due to the same incompatibility with your new handling. It generally crashed the machine. Test case is assemble the utility using revision [966fc8f14s16](https://hg.pushbx.org/ecm/keephook/rev/966fc8f14a16) and try to install it (just `keephook` as a command). The 2019 dosemus work fine, the [newer 2021 dosemu](https://github.com/dosemu2/dosemu2/commit/65853ec95) crashes instead.

Studying the code I already came up with [a workaround](https://hg.pushbx.org/ecm/keephook/rev/b71ca83f8da3) that works fine on 2021 dosemu2 and should not cause any problems for any other system. Using the applications with this commit applied makes them work on both dosemu2 versions I'm keeping around currently.

I also had another problem that only occurred because of upgrading dosemu2, which was about [the amount of system drives](https://github.com/dosemu2/dosemu2/discussions/1456).

**Reply by stsp** [reply]
What exactly the problem is?
4C does not put an extra frame.
What functions should stop
putting an extra frame?

**Reply by ecm-pushbx** [reply]
All of them, but particularly function 48h and 33h were the ones I happened to use. Refer to https://hg.pushbx.org/ecm/keephook/file/966fc8f14a16/transien.asm#l1794 for this. The stack pointer initialised by function 48h in `word [OLDPSP:2Eh]` was patched into the new PSP's field.

**Reply by stsp** [reply]
Please explain the scenario in details.
So you have some TSR that hooks int21
and patches PSP:2e, right? Then it passes
the calls down to DOS, right?
Your TSR does not get any extra iret
frame but DOS does, right?
Please describe all the events, one
by one.

**Reply by ecm-pushbx** [reply]
No, this isn't about any resident TSR. It's about process relocation. That the process relocation is used in order to later install a TSR is incidental. (And most of my TSRs do not hook interrupt 21h.)

To lay out what happens, consider the "normal" case: An application calls 21.4B00 to execute a child process. The address of the parent's user stack frame (int 21h frame stored on user stack) is saved in the `dword [2Eh]` of **the parent process**.

DOS doesn't return to the parent immediately but rather builds a new process, with a new stack. (On DOS calls with the child process active, the new stack is saved in **the child's** `dword [2Eh]` but this is of no concern to us.) Once the child terminates, the interrupt vectors 22h, 23h, and 24h are restored from the fields in the child process. The interrupt 22h vector is used as the Parent Return Address. The parent process field of the child process is used to make the parent process active again. Then the child process can be freed, its MCBs released and file handles closed (except for 21.31 TSR or when the child process is self-parented).

Now the way DOS knows what stack to resume to, is it uses the parent process's `dword [2Eh]`, expecting it to point to a user stack frame created by the 21.4B00 call. (It may clear the Carry Flag in that stack frame either earlier in 21.4B00 or now.)

In my process relocation I first set up the stack in parts of my process memory no longer used by me. Then I shrink the process's main memory block to the needed size (including the PSP, transient, resident, and message sections). Next I set up the memory allocation strategy to Last Fit, low only, UMB link disabled. Then I allocate a new memory block the same size as the shrunken process memory block. (This interrupt 21h service 48h call is the last interrupt 21h call the original process does before I copy its `word [2Eh]` to the relocated process.) Next I copy the entire process block's contents into the relocated process memory. Then I relocate the running stack and code segment to the relocation destination. Finally the relocated to-be PSP is massaged some. Part of that is the following code:

```
	mov ds, dx		; ds = old PSP
	mov word [ 0Ah ], .terminated
	mov word [ 0Ah+2 ], ax
	mov word [ 0Eh ], i23
	mov word [ 0Eh+2 ], ax
	mov word [ 12h ], i24
	mov word [ 12h+2 ], ax	; set interrupt vectors to ours
	mov word [ 16h ], ax	; set parent PSP to the relocated one
	push word [ 2Eh ]
	pop word [ cs:2Eh ]
	mov word [ cs:2Eh+2 ], ax	; set SS:SP used by process termination
```

Note that as the comment indicates, `ds` addresses the old process, and `cs` and `ax` here address the new process. The interrupt 22h, 23h, and 24h vectors of the old (child) process are initialised to our own (relocated) handlers and to return to the new relocated (parent) process. The parent process field of the old process is set to the new process. And the stack segment at `word [2Eh + 2]` in the new (parent to be) process is initialised with the new relocated stack segment. The stack pointer is initialised from the old process's value, which has been set by the service 48h call last.

The point of all that is so that the service 4Ch call (called with `cs` and `ss` already pointing to the relocated segment) will build its user stack frame in the exact location expected by the process termination that it will "return to" by loading `ss:sp` from the (relocated) parent process's `dword [2Eh]`. So all registers and the stack are returned to the code at `.terminated` in the exact same state as they were upon entry to the 4Ch call.

As I mentioned, in some of my TSR installer processes a mismatch of stack and/or register values wasn't critical, merely sometimes setting the /J switch's flag and loading low instead of high. However, KEEPHOOK passes the resident size and the number of heap blocks to allocate on the stack, and this is likely why it fails worse if the stack or register values are changed by the 4Ch call.

**Reply by stsp** [reply]
> will build its user stack frame in the exact location expected by the process termination that it will "return to"
> by loading ss:sp from the (relocated) parent process's dword [2Eh].

OK, I realize the trick.
The new (parent's) PSP:2e is a copy
of old (child's), and it was sequenced
by the last DOS call before PSP
copy, or even by 0x26 itself.
You probably don't even need to
patch it afterwards, as it should
be the copy from the very beginning.
With that you hope that 4c call
will create the stack in the same
place, and then parent would
re-use it.
So I suppose its just this:
- Create new PSP with 0x26
- make it a parent of the current one
- 2e fields are already same
- execute 0x4c

Yes, I think such scenario should
work so I disabled the tracing by
default.

**Reply by ecm-pushbx** [reply]
> You probably don't even need to patch it afterwards, as it should be the copy from the very beginning.

The segment still needs to be patched because my last DOS call was 48h with the old stack segment still. But yes, the pointer should be correct already.

> With that you hope that 4c call will create the stack in the same place, and then parent would re-use it.

Exactly.

> So I suppose its just this:
> 
>     * Create new PSP with 0x26
> 
>     * make it a parent of the current one
> 
>     * 2e fields are already same
> 
>     * execute 0x4c

I don't actually use function 26h though, I just [copy over the old PSP](https://hg.pushbx.org/ecm/keephook/file/966fc8f14a16/transien.asm#l1737) then make sure to manually [mark the process file handles closed](https://hg.pushbx.org/ecm/keephook/file/966fc8f14a16/transien.asm#l1758) in the old process so its termination won't close them.

Anyway, thanks for the patch.

**Comment by ecm-pushbx** [comment]
Here's [the log for a debug session](https://pushbx.org/ecm/test/20211102/debug.log) of the crashing KEEPHOOK revision. The particular point of interest is that it overwrites part of an MCB even when no heap blocks are allocated to interrupt handlers yet:

```
AX=0000 BX=00C6 CX=05E9 DX=1387 SP=08CA BP=1E2F SI=00A0 DI=0098
DS=9F60 ES=1E1B SS=9E6A CS=9E6A IP=0C48 NV UP EI PL NZ AC PO CY
9E6A:0C48 F3AB              rep     stosw
-dm
PSP: 9E6A
02B7 4D 0008 0018    384 B SD
02D0 4D 02E5 0013    304 B LDEBUG
02E4 4D 02E5 1B35  108 KiB LDEBUG
1E1A 4D 9E6A 00C6    3 KiB KEEPHOOK
1EE1 4D 0000 7F87  510 KiB
9E69 5A 9E6A 0195    6 KiB KEEPHOOK
9FFF 4D 0008 2100  132 KiB SC
C100 4D 0008 0146    5 KiB SD
C247 4D 0000 0006     96 B
C24E 4D C24F 0091    2 KiB COMMAND
C2E0 4D 0000 1D1E  116 KiB
DFFF 4D 0008 1000   64 KiB SC
F000 4D 0000 0513   20 KiB
F514 4D 0000 0619   24 KiB
FB2E 4D 0000 0090    2 KiB
FBBF 5A C24F 0040   1024 B COMMAND
-tp
AX=0000 BX=00C6 CX=0000 DX=1387 SP=08CA BP=1E2F SI=00A0 DI=0C6A
DS=9F60 ES=1E1B SS=9E6A CS=9E6A IP=0C4A NV UP EI PL NZ AC PO CY
9E6A:0C4A 89F9              mov     cx, di
-
```

The resident memory block at 1E1Bh is C6h paragraphs long but the `rep stosw` wrote up to below `C6Ah`, 10 bytes past the end of the block. This is because the essentially random values on the stack after the relocating 4Ch call are not properly aligned to 16-byte boundaries. The value in `bp` is also wrong but this is less important here.
